Rename `rustc_abi::Abi` to `BackendRepr`

workingjubilee opened the pull request:
Remove the confabulation of `rustc_abi::Abi` with what "ABI" actually means by renaming it to `BackendRepr`, and rename `Abi::Aggregate` to `BackendRepr::Memory`. The type never actually represented how things are passed, as that has to have `PassMode` considered, at minimum, but rather it just is how we represented some things to the backend. This conflation arose because LLVM, the primary backend at the time, would lower certain IR forms using certain ABIs. Even that only somewhat was true, as it broke down when one ventured significantly afield of what is described by the System V AMD64 ABI either by using different architectures, ABI-modifying IR annotations, the same architecture **with different ISA extensions enabled**, or other... unexpected delights.

Unfortunately both names are still somewhat of a misnomer right now, as people have written code for years based on this misunderstanding. Still, their original names are even moreso, and for better or worse, this backend code hasn't received as much maintenance as the rest of the compiler, lately. Actually arriving at a correct end-state will simply require us to disentangle a lot of code in order to fix, much of it pointlessly repeated in several places. Thus this is not an "actual fix", just a way to deflect further misunderstandings.

<!-- homu-ignore:start -->
- Acts on long-ago remarks from @eddyb as narrated by @RalfJung in https://github.com/rust-lang/compiler-team/issues/672#issuecomment-1751624436
- Part of https://github.com/rust-lang/rust/issues/119183

r? @jieyouxu
<!-- homu-ignore:end -->

compiler-errors commented:
Hate to bikeshed but why is this called "Ir"? We got lots of IRs in the compiler, and the impression I get is that type name is overly vague. It's like naming a type "Kind" with no qualifiers 😃 

LayoutForm? LayoutKind? idk

compiler-errors commented on compiler/rustc_abi/src/layout.rs:
```diff
@@ -28,7 +28,7 @@ where
     F: Deref<Target = &'a LayoutS<FieldIdx, VariantIdx>> + fmt::Debug,
 {
-    let uninhabited = fields.iter().any(|f| f.abi.is_uninhabited());
+    let uninhabited = fields.iter().any(|f| f.ir_form.is_uninhabited());
```
```suggestion
    let uninhabited = fields.iter().any(|f| f.is_uninhabited());
```

compiler-errors reviewed (commented):
> At the same time, reduce the need of some backends of rustc to snuffle around the internals of the Layout types by exposing another accessor, LayoutS::is_uninhabited.

Seems like there's quite a few callsites that are still doing `layout.ir_form.is_uninhabited()` rather than going thru the new accessor. I'd also be happy to quickly review that set of changes pulled out of this PR in particular, since it seems somewhat orthogonal to this rename.

Similarly, there are some callsites that call `layout.ir_form.other_method()` for other methods like `is_sized()` that could probably just be turned into direct calls on the layout type?

workingjubilee commented:
Because that is all it was ever supposed to be.
- It does not actually meaningfully determine layout: there are many scalars that might be represented identically to aggregates and vice versa.
- Likewise, it does not actually determine the calling convention, as there are many IR annotations that might be added via other routes, some of which have much more of an impact.

It primarily was intended to distinguish between "handle as an SSA value (scalar or vector), literally written using the syntactic forms that LLVMIR gives to SSA values, or the IR repr, or interact with it as a pointer to memory".

jieyouxu commented:
I think I like LayoutForm more than IrForm, for the same reasoning -- we already have *several* IRs in the compiler.

compiler-errors commented:
> Because that is all it was ever supposed to be.

Cool, but I still think it needs a name that clearly denotes that it is a component of computations that are primarily concerning codegen and layout. Right now that name does not do that at all. "IR" means something completely different to the bulk of the compiler than what you seem to make it mean here.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11546768279/job/32135651248) [(plain)](https://github.com/rust-lang/rust/commit/ef679043aaecefc6e01bca599bed0d3cdf55bbb5/checks/32135651248/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 47)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 50)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 63)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
---- [ui] tests/ui/type/pattern_types/range_patterns.rs stdout ----
Saved the actual stderr to "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type/pattern_types/range_patterns/range_patterns.stderr"
diff of stderr:

124                            abi: Align(1 bytes),
125                            pref: $SOME_ALIGN,
-                        abi: Aggregate {
+                        abi: Memory {
128                            sized: true,
129                        },
129                        },
130                        fields: Arbitrary {

232                            abi: Align(1 bytes),
233                            pref: $SOME_ALIGN,
-                        abi: Aggregate {
+                        abi: Memory {
236                            sized: true,
237                        },
---
To only update this specific test, also pass `--test-args type/pattern_types/range_patterns.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/type/pattern_types/range_patterns.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type/pattern_types/range_patterns" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type/pattern_types/range_patterns/auxiliary"
--- stderr -------------------------------
--- stderr -------------------------------
error: layout_of(NonZero<u32>) = Layout {
           size: Size(4 bytes),
           align: AbiAndPrefAlign {
               abi: Align(4 bytes),
               pref: Align(8 bytes),
           abi: Scalar(
               Initialized {
                   value: Int(
                       I32,
---
           unadjusted_abi_align: Align(4 bytes),
       }
  --> /checkout/tests/ui/type/pattern_types/range_patterns.rs:11:1
   |
LL | type X = std::num::NonZeroU32; //~ ERROR layout_of


error: layout_of((u32) is 1..=) = Layout {
           size: Size(4 bytes),
           align: AbiAndPrefAlign {
               abi: Align(4 bytes),
               pref: Align(4 bytes),
           abi: Scalar(
               Initialized {
                   value: Int(
                       I32,
---
           unadjusted_abi_align: Align(4 bytes),
       }
  --> /checkout/tests/ui/type/pattern_types/range_patterns.rs:13:1
   |
LL | type Y = pattern_type!(u32 is 1..); //~ ERROR layout_of


error: layout_of(Option<(u32) is 1..=>) = Layout {
           size: Size(4 bytes),
           align: AbiAndPrefAlign {
               abi: Align(4 bytes),
               pref: Align(8 bytes),
           abi: Scalar(
               Initialized {
                   value: Int(
                       I32,
                       I32,
                       false,
                   ),
                   valid_range: (..=0) | (1..),
           ),
           fields: Arbitrary {
               offsets: [
                   Size(0 bytes),
---
                   value: Int(
                       I32,
                       false,
                   ),
                   valid_range: (..=0) | (1..),
               tag_encoding: Niche {
                   untagged_variant: 1,
                   niche_variants: 0..=0,
                   niche_start: 0,
                   niche_start: 0,
               },
               tag_field: 0,
               variants: [
                   Layout {
                       size: Size(0 bytes),
                       align: AbiAndPrefAlign {
                           abi: Align(1 bytes),
                           pref: Align(8 bytes),
                       abi: Memory {
                           sized: true,
                       },
                       fields: Arbitrary {
---
                       max_repr_align: None,
                       unadjusted_abi_align: Align(1 bytes),
                   },
                   Layout {
                       size: Size(4 bytes),
                       align: AbiAndPrefAlign {
                           abi: Align(4 bytes),
                           pref: Align(8 bytes),
                       abi: Scalar(
                           Initialized {
                               value: Int(
                                   I32,
---
           unadjusted_abi_align: Align(4 bytes),
       }
  --> /checkout/tests/ui/type/pattern_types/range_patterns.rs:15:1
   |
LL | type Z = Option<pattern_type!(u32 is 1..)>; //~ ERROR layout_of


error: layout_of(Option<NonZero<u32>>) = Layout {
           size: Size(4 bytes),
           align: AbiAndPrefAlign {
               abi: Align(4 bytes),
               pref: Align(8 bytes),
           abi: Scalar(
               Initialized {
                   value: Int(
                       I32,
                       I32,
                       false,
                   ),
                   valid_range: (..=0) | (1..),
           ),
           fields: Arbitrary {
               offsets: [
                   Size(0 bytes),
---
                   value: Int(
                       I32,
                       false,
                   ),
                   valid_range: (..=0) | (1..),
               tag_encoding: Niche {
                   untagged_variant: 1,
                   niche_variants: 0..=0,
                   niche_start: 0,
                   niche_start: 0,
               },
               tag_field: 0,
               variants: [
                   Layout {
                       size: Size(0 bytes),
                       align: AbiAndPrefAlign {
                           abi: Align(1 bytes),
                           pref: Align(8 bytes),
                       abi: Memory {
                           sized: true,
                       },
                       fields: Arbitrary {
---
                       max_repr_align: None,
                       unadjusted_abi_align: Align(1 bytes),
                   },
                   Layout {
                       size: Size(4 bytes),
                       align: AbiAndPrefAlign {
                           abi: Align(4 bytes),
                           pref: Align(8 bytes),
                       abi: Scalar(
                           Initialized {
                               value: Int(
                                   I32,
---
           unadjusted_abi_align: Align(4 bytes),
       }
  --> /checkout/tests/ui/type/pattern_types/range_patterns.rs:17:1
   |
LL | type A = Option<std::num::NonZeroU32>; //~ ERROR layout_of


error: layout_of(NonZeroU32New) = Layout {
           size: Size(4 bytes),
           align: AbiAndPrefAlign {
               abi: Align(4 bytes),
               pref: Align(8 bytes),
           abi: Scalar(
               Initialized {
                   value: Int(
                       I32,
---
           unadjusted_abi_align: Align(4 bytes),
       }
  --> /checkout/tests/ui/type/pattern_types/range_patterns.rs:19:1
   |
LL | struct NonZeroU32New(pattern_type!(u32 is 1..)); //~ ERROR layout_of

error: aborting due to 5 previous errors
------------------------------------------

```

</details>

workingjubilee commented:
Happy to crank the name generator in my head, just wanted to clarify that going to "LayoutSomething" will add to the confusion.

workingjubilee commented:
SsaForm or whatever

jieyouxu commented:
That sounds reasonable to me.

jieyouxu commented:
r? @compiler-errors since you're already reviewing it

workingjubilee commented:
Yeah I got the idea of "wait... accessors exist" only midway through this.

compiler-errors commented:
I'm happy with `SsaForm` or something slightly more descriptive in that direction, but maybe let's wait a few hours (i.e. im gonna go sleep now) to see if anyone else on the backend side has stronger opinions than me.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11547177948/job/32136660212) [(plain)](https://github.com/rust-lang/rust/commit/bd156d72569e55758fc1ecc23045f62a125a77db/checks/32136660212/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 50)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 63)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 46)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
---- [ui] tests/ui/type/pattern_types/range_patterns.rs stdout ----
Saved the actual stderr to "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type/pattern_types/range_patterns/range_patterns.stderr"
diff of stderr:

124                            abi: Align(1 bytes),
125                            pref: $SOME_ALIGN,
-                        abi: Aggregate {
+                        abi: Memory {
128                            sized: true,
129                        },
129                        },
130                        fields: Arbitrary {

232                            abi: Align(1 bytes),
233                            pref: $SOME_ALIGN,
-                        abi: Aggregate {
+                        abi: Memory {
236                            sized: true,
237                        },
---
To only update this specific test, also pass `--test-args type/pattern_types/range_patterns.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/type/pattern_types/range_patterns.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type/pattern_types/range_patterns" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type/pattern_types/range_patterns/auxiliary"
--- stderr -------------------------------
--- stderr -------------------------------
error: layout_of(NonZero<u32>) = Layout {
           size: Size(4 bytes),
           align: AbiAndPrefAlign {
               abi: Align(4 bytes),
               pref: Align(8 bytes),
           abi: Scalar(
               Initialized {
                   value: Int(
                       I32,
---
           unadjusted_abi_align: Align(4 bytes),
       }
  --> /checkout/tests/ui/type/pattern_types/range_patterns.rs:11:1
   |
LL | type X = std::num::NonZeroU32; //~ ERROR layout_of


error: layout_of((u32) is 1..=) = Layout {
           size: Size(4 bytes),
           align: AbiAndPrefAlign {
               abi: Align(4 bytes),
               pref: Align(4 bytes),
           abi: Scalar(
               Initialized {
                   value: Int(
                       I32,
---
           unadjusted_abi_align: Align(4 bytes),
       }
  --> /checkout/tests/ui/type/pattern_types/range_patterns.rs:13:1
   |
LL | type Y = pattern_type!(u32 is 1..); //~ ERROR layout_of


error: layout_of(Option<(u32) is 1..=>) = Layout {
           size: Size(4 bytes),
           align: AbiAndPrefAlign {
               abi: Align(4 bytes),
               pref: Align(8 bytes),
           abi: Scalar(
               Initialized {
                   value: Int(
                       I32,
                       I32,
                       false,
                   ),
                   valid_range: (..=0) | (1..),
           ),
           fields: Arbitrary {
               offsets: [
                   Size(0 bytes),
---
                   value: Int(
                       I32,
                       false,
                   ),
                   valid_range: (..=0) | (1..),
               tag_encoding: Niche {
                   untagged_variant: 1,
                   niche_variants: 0..=0,
                   niche_start: 0,
                   niche_start: 0,
               },
               tag_field: 0,
               variants: [
                   Layout {
                       size: Size(0 bytes),
                       align: AbiAndPrefAlign {
                           abi: Align(1 bytes),
                           pref: Align(8 bytes),
                       abi: Memory {
                           sized: true,
                       },
                       fields: Arbitrary {
---
                       max_repr_align: None,
                       unadjusted_abi_align: Align(1 bytes),
                   },
                   Layout {
                       size: Size(4 bytes),
                       align: AbiAndPrefAlign {
                           abi: Align(4 bytes),
                           pref: Align(8 bytes),
                       abi: Scalar(
                           Initialized {
                               value: Int(
                                   I32,
---
           unadjusted_abi_align: Align(4 bytes),
       }
  --> /checkout/tests/ui/type/pattern_types/range_patterns.rs:15:1
   |
LL | type Z = Option<pattern_type!(u32 is 1..)>; //~ ERROR layout_of


error: layout_of(Option<NonZero<u32>>) = Layout {
           size: Size(4 bytes),
           align: AbiAndPrefAlign {
               abi: Align(4 bytes),
               pref: Align(8 bytes),
           abi: Scalar(
               Initialized {
                   value: Int(
                       I32,
                       I32,
                       false,
                   ),
                   valid_range: (..=0) | (1..),
           ),
           fields: Arbitrary {
               offsets: [
                   Size(0 bytes),
---
                   value: Int(
                       I32,
                       false,
                   ),
                   valid_range: (..=0) | (1..),
               tag_encoding: Niche {
                   untagged_variant: 1,
                   niche_variants: 0..=0,
                   niche_start: 0,
                   niche_start: 0,
               },
               tag_field: 0,
               variants: [
                   Layout {
                       size: Size(0 bytes),
                       align: AbiAndPrefAlign {
                           abi: Align(1 bytes),
                           pref: Align(8 bytes),
                       abi: Memory {
                           sized: true,
                       },
                       fields: Arbitrary {
---
                       max_repr_align: None,
                       unadjusted_abi_align: Align(1 bytes),
                   },
                   Layout {
                       size: Size(4 bytes),
                       align: AbiAndPrefAlign {
                           abi: Align(4 bytes),
                           pref: Align(8 bytes),
                       abi: Scalar(
                           Initialized {
                               value: Int(
                                   I32,
---
           unadjusted_abi_align: Align(4 bytes),
       }
  --> /checkout/tests/ui/type/pattern_types/range_patterns.rs:17:1
   |
LL | type A = Option<std::num::NonZeroU32>; //~ ERROR layout_of


error: layout_of(NonZeroU32New) = Layout {
           size: Size(4 bytes),
           align: AbiAndPrefAlign {
               abi: Align(4 bytes),
               pref: Align(8 bytes),
           abi: Scalar(
               Initialized {
                   value: Int(
                       I32,
---
           unadjusted_abi_align: Align(4 bytes),
       }
  --> /checkout/tests/ui/type/pattern_types/range_patterns.rs:19:1
   |
LL | struct NonZeroU32New(pattern_type!(u32 is 1..)); //~ ERROR layout_of

error: aborting due to 5 previous errors
------------------------------------------

```

</details>

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11547904460/job/32138511668) [(plain)](https://github.com/rust-lang/rust/commit/23a617502141eb4bd373e2e99de878ff30a7b938/checks/32138511668/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 3.153 Building wheels for collected packages: reuse
#13 3.154   Building wheel for reuse (pyproject.toml): started
#13 3.396   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.397   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#13 3.398   Stored in directory: /tmp/pip-ephem-wheel-cache-u461mmyh/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.400 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.799 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.799 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.374 Collecting virtualenv
#13 4.374 Collecting virtualenv
#13 4.443   Downloading virtualenv-20.27.0-py3-none-any.whl (3.1 MB)
#13 4.676      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 13.5 MB/s eta 0:00:00
#13 4.737 Collecting platformdirs<5,>=3.9.1
#13 4.748   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.773 Collecting distlib<1,>=0.3.7
#13 4.805      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 24.2 MB/s eta 0:00:00
#13 4.846 Collecting filelock<4,>=3.12.2
#13 4.857   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.857   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.938 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.116 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.27.0
#13 DONE 5.2s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      206784 kB
DirectMap2M:     8181760 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://static.rust-lang.org/dist/2024-10-16/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/2024-10-16/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt
---
fmt check
Diff in /checkout/compiler/rustc_ty_utils/src/layout/invariant.rs:262:
                 )
             }
             // Skip empty variants.
-            if variant.size == Size::ZERO
-                || variant.fields.count() == 0
-                || variant.is_uninhabited()
+            if variant.size == Size::ZERO || variant.fields.count() == 0 || variant.is_uninhabited()
                 // These are never actually accessed anyway, so we can skip the coherence check
                 // for them. They also fail that check, since they have
                 // for them. They also fail that check, since they have
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_query_system/src/dep_graph/debug.rs" "/checkout/compiler/rustc_query_system/src/dep_graph/edges.rs" "/checkout/compiler/rustc_query_system/src/dep_graph/serialized.rs" "/checkout/compiler/rustc_query_system/src/dep_graph/graph.rs" "/checkout/compiler/rustc_query_system/src/dep_graph/query.rs" "/checkout/compiler/rustc_query_system/src/dep_graph/mod.rs" "/checkout/compiler/rustc_query_system/src/query/config.rs" "/checkout/compiler/rustc_query_system/src/query/caches.rs" "/checkout/compiler/rustc_query_system/src/query/plumbing.rs" "/checkout/compiler/rustc_query_system/src/query/job.rs" "/checkout/compiler/rustc_query_system/src/query/mod.rs" "/checkout/compiler/rustc_query_system/src/ich/impls_syntax.rs" "/checkout/compiler/rustc_query_system/src/ich/hcx.rs" "/checkout/compiler/rustc_query_system/src/ich/mod.rs" "/checkout/compiler/rustc_query_system/src/values.rs" "/checkout/compiler/rustc_query_system/src/lib.rs" "/checkout/compiler/rustc_query_system/src/error.rs" "/checkout/compiler/rustc_hir/src/stable_hash_impls.rs" "/checkout/compiler/rustc_hir/src/weak_lang_items.rs" "/checkout/compiler/rustc_hir/src/def_path_hash_map.rs" "/checkout/compiler/rustc_hir/src/definitions.rs" "/checkout/compiler/rustc_hir/src/tests.rs" "/checkout/compiler/rustc_hir/src/pat_util.rs" "/checkout/compiler/rustc_hir/src/hir_id.rs" "/checkout/compiler/rustc_hir/src/lang_items.rs" "/checkout/compiler/rustc_hir/src/lib.rs" "/checkout/compiler/rustc_hir/src/intravisit.rs" "/checkout/compiler/rustc_hir/src/target.rs" "/checkout/compiler/rustc_hir/src/diagnostic_items.rs" "/checkout/compiler/rustc_hir/src/hir.rs" "/checkout/compiler/rustc_hir/src/def.rs" "/checkout/compiler/rustc_hir/src/arena.rs" "/checkout/compiler/rustc_lint_defs/src/builtin.rs" "/checkout/compiler/rustc_lint_defs/src/lib.rs" "/checkout/compiler/stable_mir/src/ty.rs" "/checkout/compiler/stable_mir/src/abi.rs" "/checkout/compiler/stable_mir/src/mir.rs" "/checkout/compiler/stable_mir/src/crate_def.rs" "/checkout/compiler/stable_mir/src/compiler_interface.rs" "/checkout/compiler/stable_mir/src/lib.rs" "/checkout/compiler/stable_mir/src/visitor.rs" "/checkout/compiler/stable_mir/src/target.rs" "/checkout/compiler/stable_mir/src/mir/mono.rs" "/checkout/compiler/stable_mir/src/mir/body.rs" "/checkout/compiler/stable_mir/src/mir/alloc.rs" "/checkout/compiler/stable_mir/src/mir/visit.rs" "/checkout/compiler/stable_mir/src/mir/pretty.rs" "/checkout/compiler/stable_mir/src/error.rs" "/checkout/compiler/rustc_ty_utils/src/ty.rs" "/checkout/compiler/rustc_ty_utils/src/layout/invariant.rs" "/checkout/compiler/rustc_ty_utils/src/sig_types.rs" "/checkout/compiler/rustc_ty_utils/src/assoc.rs" "/checkout/compiler/rustc_ty_utils/src/abi.rs" "/checkout/compiler/rustc_ty_utils/src/opaque_types.rs" "/checkout/compiler/rustc_ty_utils/src/errors.rs" "/checkout/compiler/rustc_ty_utils/src/consts.rs" "/checkout/compiler/rustc_ty_utils/src/common_traits.rs" "/checkout/compiler/rustc_ty_utils/src/instance.rs" "/checkout/compiler/rustc_ty_utils/src/lib.rs" "/checkout/compiler/rustc_ty_utils/src/implied_bounds.rs" "/checkout/compiler/rustc_ty_utils/src/structural_match.rs" "/checkout/compiler/rustc_ty_utils/src/representability.rs" "/checkout/compiler/rustc_ty_utils/src/layout.rs" "/checkout/compiler/rustc_query_system/src/dep_graph/dep_node.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Mon Oct 28 05:33:07 UTC 2024
  network time: Mon, 28 Oct 2024 05:33:07 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

RalfJung commented:
`SsaForm` makes little sense to me since a codegen backend doesn't have to use SSA -- for instance, Miri uses this to inform the way it represents local variables, but there's no SSA involved.

In the MCP I suggested names like `StorageClass` or `StorageKind`. We could also use `StorageForm`. To me, "storage" makes sense since it is really a detail about how the rest of the compiler is supposed to store values of this type -- though it is always correct to ignore that and store it in-memory anyway.

workingjubilee commented:
"Storage" just sounds like a synonym for "Layout" though.

workingjubilee commented:
The unfortunate problem here is that this type currently _does not answer a question_, but we can't simply discard it without significantly more work.

To call it "Storage" implies it would answer the question of "how will this value be stored?"

But it does not.

RalfJung commented:
Not at all? "layout" for us has the fairly clear meaning of being about field offsets, and related properties. It's like the layout of a page or website. "storage" has a completely different meaning than that? I don't follow at all TBH.

RalfJung commented:
The type is a hint, answering the question of how values of this type shall be stored by backends if there are no other constraints that force it to be stored in-memory.

workingjubilee commented:
> The type is a hint, answering the question of how values of this type shall be stored by backends if there are no other constraints that force it to be stored in-memory.

`PoliteAbiSuggestionIfItIsntTooMuchTrouble`, then???

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11548086591/job/32138955178) [(plain)](https://github.com/rust-lang/rust/commit/9b440837a721d5af97e775e4ac1aee600c56eecf/checks/32138955178/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:c32c805632780b5c1de330e3f44561b336c2efe163bc0990acb392390157a8e1d9f855d75914a239aa40c49d77f4a837247d05d2f8d46f554b98e1f46712a3e3:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
---- [ui] tests/ui/type/pattern_types/range_patterns.rs stdout ----
Saved the actual stderr to "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type/pattern_types/range_patterns/range_patterns.stderr"
diff of stderr:

124                            abi: Align(1 bytes),
125                            pref: $SOME_ALIGN,
-                        abi: Aggregate {
+                        abi: Memory {
128                            sized: true,
129                        },
129                        },
130                        fields: Arbitrary {

232                            abi: Align(1 bytes),
233                            pref: $SOME_ALIGN,
-                        abi: Aggregate {
+                        abi: Memory {
236                            sized: true,
237                        },
---
To only update this specific test, also pass `--test-args type/pattern_types/range_patterns.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/type/pattern_types/range_patterns.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type/pattern_types/range_patterns" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type/pattern_types/range_patterns/auxiliary"
--- stderr -------------------------------
--- stderr -------------------------------
error: layout_of(NonZero<u32>) = Layout {
           size: Size(4 bytes),
           align: AbiAndPrefAlign {
               abi: Align(4 bytes),
               pref: Align(8 bytes),
           abi: Scalar(
               Initialized {
                   value: Int(
                       I32,
---
           unadjusted_abi_align: Align(4 bytes),
       }
  --> /checkout/tests/ui/type/pattern_types/range_patterns.rs:11:1
   |
LL | type X = std::num::NonZeroU32; //~ ERROR layout_of


error: layout_of((u32) is 1..=) = Layout {
           size: Size(4 bytes),
           align: AbiAndPrefAlign {
               abi: Align(4 bytes),
               pref: Align(4 bytes),
           abi: Scalar(
               Initialized {
                   value: Int(
                       I32,
---
           unadjusted_abi_align: Align(4 bytes),
       }
  --> /checkout/tests/ui/type/pattern_types/range_patterns.rs:13:1
   |
LL | type Y = pattern_type!(u32 is 1..); //~ ERROR layout_of


error: layout_of(Option<(u32) is 1..=>) = Layout {
           size: Size(4 bytes),
           align: AbiAndPrefAlign {
               abi: Align(4 bytes),
               pref: Align(8 bytes),
           abi: Scalar(
               Initialized {
                   value: Int(
                       I32,
                       I32,
                       false,
                   ),
                   valid_range: (..=0) | (1..),
           ),
           fields: Arbitrary {
               offsets: [
                   Size(0 bytes),
---
                   value: Int(
                       I32,
                       false,
                   ),
                   valid_range: (..=0) | (1..),
               tag_encoding: Niche {
                   untagged_variant: 1,
                   niche_variants: 0..=0,
                   niche_start: 0,
                   niche_start: 0,
               },
               tag_field: 0,
               variants: [
                   Layout {
                       size: Size(0 bytes),
                       align: AbiAndPrefAlign {
                           abi: Align(1 bytes),
                           pref: Align(8 bytes),
                       abi: Memory {
                           sized: true,
                       },
                       fields: Arbitrary {
---
                       max_repr_align: None,
                       unadjusted_abi_align: Align(1 bytes),
                   },
                   Layout {
                       size: Size(4 bytes),
                       align: AbiAndPrefAlign {
                           abi: Align(4 bytes),
                           pref: Align(8 bytes),
                       abi: Scalar(
                           Initialized {
                               value: Int(
                                   I32,
---
           unadjusted_abi_align: Align(4 bytes),
       }
  --> /checkout/tests/ui/type/pattern_types/range_patterns.rs:15:1
   |
LL | type Z = Option<pattern_type!(u32 is 1..)>; //~ ERROR layout_of


error: layout_of(Option<NonZero<u32>>) = Layout {
           size: Size(4 bytes),
           align: AbiAndPrefAlign {
               abi: Align(4 bytes),
               pref: Align(8 bytes),
           abi: Scalar(
               Initialized {
                   value: Int(
                       I32,
                       I32,
                       false,
                   ),
                   valid_range: (..=0) | (1..),
           ),
           fields: Arbitrary {
               offsets: [
                   Size(0 bytes),
---
                   value: Int(
                       I32,
                       false,
                   ),
                   valid_range: (..=0) | (1..),
               tag_encoding: Niche {
                   untagged_variant: 1,
                   niche_variants: 0..=0,
                   niche_start: 0,
                   niche_start: 0,
               },
               tag_field: 0,
               variants: [
                   Layout {
                       size: Size(0 bytes),
                       align: AbiAndPrefAlign {
                           abi: Align(1 bytes),
                           pref: Align(8 bytes),
                       abi: Memory {
                           sized: true,
                       },
                       fields: Arbitrary {
---
                       max_repr_align: None,
                       unadjusted_abi_align: Align(1 bytes),
                   },
                   Layout {
                       size: Size(4 bytes),
                       align: AbiAndPrefAlign {
                           abi: Align(4 bytes),
                           pref: Align(8 bytes),
                       abi: Scalar(
                           Initialized {
                               value: Int(
                                   I32,
---
           unadjusted_abi_align: Align(4 bytes),
       }
  --> /checkout/tests/ui/type/pattern_types/range_patterns.rs:17:1
   |
LL | type A = Option<std::num::NonZeroU32>; //~ ERROR layout_of


error: layout_of(NonZeroU32New) = Layout {
           size: Size(4 bytes),
           align: AbiAndPrefAlign {
               abi: Align(4 bytes),
               pref: Align(8 bytes),
           abi: Scalar(
               Initialized {
                   value: Int(
                       I32,
---
           unadjusted_abi_align: Align(4 bytes),
       }
  --> /checkout/tests/ui/type/pattern_types/range_patterns.rs:19:1
   |
LL | struct NonZeroU32New(pattern_type!(u32 is 1..)); //~ ERROR layout_of

error: aborting due to 5 previous errors
------------------------------------------

```

</details>

RalfJung commented:
Sure. ;)

Maybe `StorageHint`? Though that name would be aspirational, since currently we *do* also use this type to inform the actual call ABI.

workingjubilee commented:
Hmm. `SomethingHint` does seem the way to go! Currently weighing `StorageHint` versus `AbiHint` (less of a change, annoyingly, but possibly more accurate?) or `ReprHint` (more ambiguous, which could be good?).

RalfJung commented:
It's currently not just a hint though. For instance it is also the authority on whether this is an unsized type. So I feel like further refactoring would be required before we can use a name like that.

bjorn3 commented:
rustc_abi::Abi does have significant impact on the calling convention, though it indeed doesn't determine the calling convention on it's own. For example the rust abi computation does take rustc_abi::Abi into account when determining the pass mode to use. Also for example PassMode::Direct is only valid when the type is Abi::Scalar or Abi::Vector and similar for PassMode::Pair and Abi::ScalarPair. If you have Abi::Aggregate you have to use PassMode::Cast for passing it in register. I'm not sure what the best name is, but I do completely agree that it should be renamed to something other than rustc_abi::Abi to reduce confusion.

bors commented:
:umbrella: The latest upstream changes (presumably #132262) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

workingjubilee commented:
My intention was always to make a PR to change the name and later change the code to more accurately reflect the name, yes.

I realize the result will be code that reads incorrectly in a few spots, but that also is intentional, so that the code is provocatively weird, rather than being "confidently wrong". Then, when I fix up the code later and pull a reviewer, rather than needing to explain that the `Abi` type's meaning is bogus, and that I am incrementally changing towards a certain direction, and maybe relitigating that change, the resulting diffs read as obviously reducing the incongruity. Maybe I should also add a `FIXME` to every single field access to `LayoutData` too, I guess.

Obviously if the reviewer is one of the people currently reading this it would go a bit easier,  but I prefer to aim for allowing people that don't have perfect context to review the code.

I don't really want to use something too definite like `Storage`, as I don't know if it will actually describe `Storage`, nor if I _want_ it to, yet. I'm going to be doing a lot to change a lot of adjacent details. So if it takes on a definite form, I'd rather it mostly signal the (original) intention as being about how we represent something to the codegen backend... That suggests another 🎨 choice: `CodegenRepr` maybe.

RalfJung commented:
I'd prefer "backend" over "codegen" since we have a backend that does not generate code. ;)

workingjubilee commented:
That would also be fine, I think.

compiler-errors commented:
So `BackendRepr` it is? :D

RalfJung commented:
We can always rename again if we find a better name. ;)

bors commented:
:umbrella: The latest upstream changes (presumably #132277) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

workingjubilee committed:
compiler: `rustc_abi::Abi` => `BackendRepr`

The initial naming of "Abi" was an awful mistake, conveying wrong ideas
about how psABIs worked and even more about what the enum meant.
It was only meant to represent the way the value would be described to
a codegen backend as it was lowered to that intermediate representation.
It was never meant to mean anything about the actual psABI handling!
The conflation is because LLVM typically will associate a certain form
with a certain ABI, but even that does not hold when the special cases
that actually exist arise, plus the IR annotations that modify the ABI.

Reframe `rustc_abi::Abi` as the `BackendRepr` of the type, and rename
`BackendRepr::Aggregate` as `BackendRepr::Memory`. Unfortunately, due to
the persistent misunderstandings, this too is now incorrect:
- Scattered ABI-relevant code is entangled with BackendRepr
- We do not always pre-compute a correct BackendRepr that reflects how
  we "actually" want this value to be handled, so we leave the backend
  interface to also inject various special-cases here
- In some cases `BackendRepr::Memory` is a "real" aggregate, but in
  others it is in fact using memory, and in some cases it is a scalar!

Our rustc-to-backend lowering code handles this sort of thing right now.
That will eventually be addressed by lifting duplicated lowering code
to either rustc_codegen_ssa or rustc_target as appropriate.

workingjubilee committed:
cg_clif: `rustc_abi::Abi` => `BackendRepr`

workingjubilee committed:
cg_gcc: `rustc_abi::Abi` => `BackendRepr`

workingjubilee committed:
miri: `rustc_abi::Abi` => `BackendRepr`

workingjubilee committed:
rust-analyzer: `rustc_abi::Abi` => `BackendRepr`

workingjubilee commented:
As I was working through rebasing and updating the commits with the new decision, I kept abbreviating `BackendRepr` to `BackRepr`, so I just shrugged and went with it.

compiler-errors commented on compiler/rustc_abi/src/layout/ty.rs:
```diff
@@ -83,8 +83,8 @@ impl<'a> Layout<'a> {
 
-    pub fn abi(self) -> Abi {
-        self.0.0.abi
+    pub fn abi(self) -> BackendRepr {
```
naming

compiler-errors commented on compiler/rustc_abi/src/lib.rs:
```diff
@@ -1686,7 +1704,7 @@ where
             size,
             align,
-            abi,
+            backend_repr: abi,
```
rename the `abi` local here?

compiler-errors commented on compiler/rustc_codegen_cranelift/src/value_and_place.rs:
```diff
@@ -646,23 +650,28 @@ impl<'tcx> CPlace<'tcx> {
-                        _ => bug!("Non ScalarPair abi {:?} for ByValPair CValue", dst_layout.abi),
+                        _ => {
+                            bug!(
+                                "Non ScalarPair abi {:?} for ByValPair CValue",
```
```suggestion
                                "Non ScalarPair backend repr {:?} for ByValPair CValue",
```

compiler-errors commented on compiler/rustc_ty_utils/src/layout.rs:
```diff
@@ -330,17 +333,17 @@ fn layout_of_uncached<'tcx>(
                 variants: Variants::Single { index: FIRST_VARIANT },
                 fields: FieldsShape::Array { stride: element.size, count },
-                abi,
+                backend_repr: abi,
```
maybe rename this abi local

compiler-errors reviewed (approved):
Some minor nits, take them or leave them. r=me otherwise.

workingjubilee commented:
Dropped the random cleanup commit because it has no more path-dependence on this PR (it used to be bigger and involve the Cargo.tomls in the tree).

workingjubilee commented:
Minor rewording of the initial commit message.

This seems likely to conflict, given it touches most of our subtrees, so

@bors r+ rollup=iffy

bors commented:
:pushpin: Commit 3c145dab3d1ec6566602d66cbe759d250d8a94ee has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee 3c145dab3d1ec6566602d66cbe759d250d8a94ee -->
<!-- homu: {"type":"Approved","sha":"3c145dab3d1ec6566602d66cbe759d250d8a94ee","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

workingjubilee commented:
...
@bors r=compiler-errors

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #132326

bors commented:
:pushpin: Commit 3c145dab3d1ec6566602d66cbe759d250d8a94ee has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 3c145dab3d1ec6566602d66cbe759d250d8a94ee -->
<!-- homu: {"type":"Approved","sha":"3c145dab3d1ec6566602d66cbe759d250d8a94ee","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

workingjubilee commented:
@bors r-

workingjubilee commented on tests/ui/abi/c-zst.other-linux.stderr:
the other variants of this went unblessed.

workingjubilee committed:
tests: Bless `rustc_abi::Abi::Aggregate` => `::Memory`

workingjubilee committed:
tests: cross-compile multi-platform ZST ABI tests

This allows them to be blessed, regardless of platform.

workingjubilee commented:
Rewrote the platform-specific ZST ABI tests to now do cross-compilation, as they each only really need a single Sized lang-item. They are now *also* blessed.

RalfJung commented on tests/ui/abi/c-zst.rs:
```diff
@@ -1,27 +1,40 @@
-//@ revisions: other other-linux x86_64-pc-windows-gnu s390x-linux sparc64-linux powerpc-linux
 //@ normalize-stderr-test: "(abi|pref|unadjusted_abi_align): Align\([1-8] bytes\)" -> "$1: $$SOME_ALIGN"
+
+// Ignore the ZST revisions
```
I don't understand the comment. What are we supposed to ignore?

RalfJung replied:
The old code had two sections,"ignore the ZST" and "pass the ZST indirectly". That structure seems to have been lost and now it's just confusing.

workingjubilee replied:
I found the previous structure confusing too, tbh.

workingjubilee replied:
@RalfJung I don't feel confident that our handling of passing the ZST by indirection is correct for the ABIs in question.

RalfJung replied:
> I found the previous structure confusing too, tbh.

There are two kinds of targets: those where ZST are ignored, and those where they are passed indirectly (which may seem nonsensical but it's what it is). It is helpful to group the revisions for the test by which category a target falls into.

workingjubilee replied:
@RalfJung I did a bit more studying to satisfy my curiosity a bit more as to exactly _why_ so many ABIs decide ZSTs consume space, and organized the revisions accordingly again, plus added some remarks as to the history.

RalfJung replied:
Thanks!

I was going to suggest a better place for the comment would be the code in rustc that implements this ABI logic. That used to be one central place but https://github.com/rust-lang/rust/commit/b1493ba5194fcc5cfe4f6315db288e4e18509110 recently moved it to the corresponding targets... which is probably a good call but makes it harder to to put a good comment in a central place.

workingjubilee replied:
Yeah, "the code may or may not be in one place" is basically the reason I've been leaving these sorts of comments by the test cases. Tests in many ways describe the spec that we're adhering to, but the implementation can be whatever... including "spread out across 100 modules".

I may find a better home for some of this text at some point, but not today.

bors commented:
:pushpin: Commit 083a362dd113da1217f7eed0e48baa2c584ac641 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 083a362dd113da1217f7eed0e48baa2c584ac641 -->
<!-- homu: {"type":"Approved","sha":"083a362dd113da1217f7eed0e48baa2c584ac641","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

workingjubilee commented:
post-hoc note: If this commit broke your custom driver, codegen backend, or other compiler extension, try to move touching on the `layout.backend_repr` to the convenience functions now available, like `LayoutData::is_uninhabited`.
